### ISSUE TITLE
Contravariance broke in 3.8 when combining spreading parameter list with conditional type

### ISSUE THREAD

**Issue by nicojs** [microsoft/TypeScript#37400]
**TypeScript Version:**  3.8.3

**Search Terms:** variance contravariance

**Description**: An Injector in [typed-inject](https://github.com/nicojs/typed-inject) is a dependency injection container that is type safe. It can only provide values for tokens it knows. A small example:

```ts
declare const fooInjector: Injector<{foo: string}>; // this injector can provide string values for the token 'foo'
const foo = fooInjector.resolve('foo'); 
// => typeof foo === 'string'
fooInjector.resolve('bar');
// => Error ts(2345) Argument of type '"bar"' is not assignable to parameter of type '"foo"'
```

It makes sense that an injector `Injector<{}>` is not assignable to `Injector<{foo: string}>`, since it cannot provide a value for token `'foo'`. This was the case in TS 3.7. However, since TS 3.8, `Injector<{}>` _is_ assignable to `Injector<{foo: string}>` 😪.

```ts
declare const rootInjector: Injector<{}>;
const fooInjector: Injector<{ foo: string}> = rootInjector;
```

**Expected behavior:** `Type 'Injector<{}>' is not assignable to type 'Injector<{ foo: string; }>'.`

**Actual behavior:** No error

**Related Issues:** Couldn't find any 🤷‍♂️

**Code**

I think I've trimmed it down to the essentials.

```ts
interface Injector<TContext> {
  injectFunction<Tokens extends (keyof TContext)[]>(todo:
    (...args: { [K in keyof Tokens]: Tokens[K] extends keyof TContext ? TContext[Tokens[K]] : never; }) => void): void;
} 

declare const rootInjector: Injector<{}>;
const fooInjector: Injector<{ foo: string}> = rootInjector;
```

<details><summary><b>Output</b></summary>

(none)

</details>

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "alwaysStrict": true,
    "esModuleInterop": true,
    "declaration": true,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "moduleResolution": 2,
    "target": "ES2017",
    "jsx": "React",
    "module": "ESNext"
  }
}
```

</details>

**Playground Link:** [Provided](https://www.typescriptlang.org/play/#code/JYOwLgpgTgZghgYwgAgJIgFYQWA9lAHgBUBhXcCADzAD5kBvAKGWVCxwDEBXEHYc4rgDWEEAGdkVSCAAmEgBQiAnrhjJS5SNQCUAbQC6NeXhm4AXMxbJ5AOjtwoAczFmGyXQGlWIZMtXrhUTF9VyJA8U99SWpROV8IFTUNCmpkAH51MhSwXTCRCI99KNcQCAA3aABuZABfbWQAXjoy3GAZbVcWtsrGGuRGRhlsABsHFARyMTBkKFxcMHR2PChXRexlgnoamh6J8WmYObWcfFXMdfxN5EPzZCmoUEdtxpm5hfOTqB6gA)
      
Simpler contravariant examples like this still work as expected.

```ts
type Action<T> = (arg: T) => void;
declare let b: Action<{}>;
declare let a: Action<{foo: number}>;
b = a
// => Error! Type 'Action<{ foo: number; }>' is not assignable to type 'Action<{}>'.
```

**Comment by RyanCavanaugh** [comment]
Tried replacing method-syntax declarations with arrow function equivalents; no effect.

@nicojs the simpler the repro the sooner we can look at this; any progress would be appreciated

**Comment by nicojs** [comment]
Thanks for the encouragement @RyanCavanaugh 💖

I've tried to trim it down further, with moderate success:

[Playground link](https://www.typescriptlang.org/play?ts=3.8.3#code/C4TwDgpgBAwg9gJwRAzmOA7AJgSwwcwBVxUAeQ+DYCAD2ABopCBlKW67FKACgGsIQcAGZNK1OgEoA2gF0AfFAC8UAN4AoKFCkBpKHij9BIljIBcTZjpls6ETgYHDRmccAD8zquyksrZqBgQAG4QCADcagC+EWp41AhCAIYAxtAAkhgAVhDJwIkARgA2EABiAK4YuTiY5GLsjABKjIRw-Bhc7HZYXHyOxnWSsgrqmtwAdBOJCPgo5vBIqOjYeEQkKLUu9UytdihyEuYNEZrIiViYhSB6WTnA5i1tKBGRarFeCSnpN7mIG150ww012yuXKlWA1QwpCa20eNg43R4hicFE2g3k3GAcHO5gyILyRVKFSqNVR-wYUBhD12+0OzygaiwOUKU2gyUwKGAUAQcDgwDxt0QuO+WIQpBUkTkEXZ7S5Ql5Ap+CGF+N+Kig8rg5k5CBWkqU3N5-JFiBiryAA)

Maybe some more clues. Performance of compiling a project with typed-inject degraded significantly with 3.8 compared to 3.7 (about 3~4 times slower). See the integration test output snippets below.

TypeScript 3.7 ([link](https://github.com/nicojs/typed-inject/runs/510916026?check_suite_focus=true))
```
  typed-inject
    ✓ dependency-graph.ts (5308ms)
    ✓ forgot-tokens-class.ts (4536ms)
    ✓ forgot-tokens-function.ts (3395ms)
    ✓ injector-too-variant.ts (3386ms)
    ✓ override-token-should-change-type.ts (3326ms)
    ✓ tokens-of-type-string.ts (3445ms)
    ✓ unknown-token.ts (3562ms)
    ✓ wrong-order-tokens.ts (3209ms)
```

TypeScript 3.8 ([link](https://github.com/nicojs/typed-inject/runs/510916026?check_suite_focus=true))

```
  typed-inject
    ✓ dependency-graph.ts (15980ms)
    ✓ forgot-tokens-class.ts (16109ms)
    ✓ forgot-tokens-function.ts (15598ms)
    1) injector-too-variant.ts
    ✓ override-token-should-change-type.ts (15747ms)
    ✓ tokens-of-type-string.ts (15607ms)
    ✓ unknown-token.ts (15563ms)
    ✓ wrong-order-tokens.ts (15932ms)
```

**Comment by nicojs** [comment]
I think I've trimmed it down to the essentials.

[playground link](https://www.typescriptlang.org/play/?ssl=8&ssc=1&pln=1&pc=1#code/JYOwLgpgTgZghgYwgAgJIgFYQWA9lAHgBUBhXcCADzAD5kBvAKGWVCxwDEBXEHYc4rgDWEEAGdkVSCAAmEgBQiAnrhjJS5SNQCUAbQC6NeXhm4AXMxbJ5AOjtwoAczFmGyXQGlWIZMtXrhUTF9VyJA8U99SWpROV8IFTUNCmpkAH51MhSwXTCRCI99KNcQCAA3aABuZABfbWQAXjoy3GAZbVcWtsrGGuRGRhlsABsHFARyMTBkKFxcMHR2PChXRexlgnoamh6J8WmYObWcfFXMdfxN5EPzZCmoUEdtxpm5hfOTqB6gA)

```ts
interface Injector<TContext> {
  injectFunction<Tokens extends (keyof TContext)[]>(todo:
    (...args: { [K in keyof Tokens]: Tokens[K] extends keyof TContext ? TContext[Tokens[K]] : never; }) => void): void;
} 

declare const rootInjector: Injector<{}>;
const fooInjector: Injector<{ foo: string}> = rootInjector; // Missing error here in TS 3.8
```

The error seems to occur when combining a _conditional type_ with _spreading parameter lists with tuples_. I'll also update the title and the inial message to better reflect this.

**Comment by nicojs** [comment]
@weswigham or @RyanCavanaugh any updates on this? Would this make it into TypeScript 4.0? 😇

**Comment by weswigham** [comment]
So I took a peek at this one today, and despite the small repro, there are a few issues going on at once.

First: When comparing instantiations of the same signature (as we do here for the `injectFunction` method), we erase generics (to `any`). This is inaccurate, and causes us to always report two instances of an `Injector` are assignable to one-another (essentially, the type parameter, since it's used mostly as a constraint, just goes away). This is a known flaw in our signature relationship checking, and has a possible fix out at #31029.

Second: Once there's a fix or workaround for that in place, this `Injector` type will fully invert it's behavior - it'll go form _accepting_ all assignments to _rejecting_ all (non-identical) assignments. The argument comparisons rely on comparing the constraint of the parameter type (`string[]`) to the still-generic parameter type (`{ [K in keyof Tokens]: Tokens[K] extends "foo" ? {foo: string}[Tokens[K]] : never; }`). _That_ will currently unconditionally return `false`, as, even when you peek through the mapped type to the element types (which we do), when we compare `string` to `Tokens[K] extends "foo" ? {foo: string}[Tokens[K]] : never;`, we're comparing against a generic conditional, which currently unconditionally returns `false`. _That_, in turn, has a _possible_ fix up at #30639.

So this small example seems to hit upon a lot of our relationship checking flaws all at once~

**Comment by nicojs** [comment]
Thanks for looking into this @weswigham 

I'm not following all the details here, but am I correct in assuming that once #30639 and #31029 are merged this issue is fixed? 👐

**Comment by nicojs** [comment]
@weswigham  #31029 is closed! 😨 Is there a follow-up planned? I guess this isn't a "Good first issue" thing?

**Comment by weswigham** [comment]
I... Dunno. I wasn't even responsible for getting #30639 merged - I'm actually mostly just idly browsing notifications while I'm conscious while in the hospital (heh). @sandersn might know if there's a larger area he's trying to merge fixes in and would better know if #31029 and updating it meets that bar for him.

**Comment by sandersn** [comment]
I closed #31029 because it had been sitting stale for a year after we decided in the design meeting that it needed some more design work. It's still worthwhile to follow up on the design meeting recommendations and see if they work out.

**Comment by nicojs** [comment]
Wow! Is it Christmas already? This bug seems to be fixed in TS 4.5. @weswigham can you confirm?

🎄

**Comment by RyanCavanaugh** [comment]
This does seem fixed 👍
